docs/design/generic.md

```diff
@@ -0,0 +1,123 @@
+
+### Source repository and cachi2 lockfile
+
+This section will describe the structure of the source repository, that will serve as an input to cachi2. The idea is to
+define a cachi2 lockfile that will specify individual artifacts to fetch along with necessary metadata - e.g. checksums.
+The format chosen for this lockfile is yaml, and will include [purl] for each
+of the fetched artifacts. This decision was made mainly because it allows for followup implementation of maven support,
+with accurate SBOM information. Alternatively, for better user experience, the download url and checksums can be specified
+separately, always resulting in a `pkg:generic` purl. Here's an example of such a lockfile with both options.
+
+```yaml
+artifacts:
+  - purl: pkg:generic/granite-model?download_url=https://huggingface.co/instructlab/granite-7b-lab/resolve/main/model-00001-of-00003.safetensors?download=true
+    target: granite-model-1.safetensors
+    checksums:
+      sha256: 07123e1f482356c415f684407a3b8723e10b2cbbc0b8fcd6282c49d37c9c1abc
+  - download_url: https://huggingface.co/instructlab/granite-7b-lab/resolve/main/model-00002-of-00003.safetensors?download=true
+    target: granite-model-2.safetensors
+    checksums:
+        sha256: 90bffe1884b84d5e255f12ff0ecbd70f2edfc877b68d612dc6fb50638b3ac17c
+```
+
+#### Specify by purl
+
+This option is provided mostly as a necessary step for the maven support, but can be used for generic artifacts as well.
+At this time, the only purl is `pkg:generic` and will result in a `pkg:generic` SBOM component.
+
+##### purl (required)
+
+At this point, the only purl type allowed would be `pkg:generic`. This is because cachi2 has no good way of verifying
+additional properties of the fetched artifact that could be included in the resulting SBOM. This should create a strong
+incentive to use this feature in the only truly necessary cases, because it will generate low-quality SBOM components,
+as compared to using other package managers provided by cachi2. Additionally, the only allowed qualifiers will be
+`download_url` and `checksums`.
+
+#### Specify by download_url and checksums
+
+This option is provided for better user experience by specifying the url and checksums separately. This option will always
+result in a `pkg:generic` purl.
+
+##### download_url (required)
+
+Specified as a string containing the download url of the artifact.
+
+##### checksums (required)
+
+Specified as a dictionary of checksum algorithms and their values. At least one cachi2-verifiable checksum must be provided.
+
+#### target (optional)
+
+This key is common for both options and providedmainly for the users convenience, so the files end up in expected locations.
+Target here means a specific subdirectory inside cachi2's output directory (likely `cachi2-output/deps/generic`).
+Special care needs to be taken to ensure there is not a conflict with other downloaded files. If not specified, filename
+of the downloaded file will be used.
```

Comment:
I'd argue that just like our packages are currently categorized based on the connection to a particular pkg manager, this should almost exclusively be `generic`, otherwise (in case of Maven or whatnot) we'll have to figure it out again based on some common pattern/relevance to some pkg ecosystem, so I question the need for this field, because currently cachi2 cannot be configured like this for any pkg manager apart from pointing it to an alternative top-level directory. I feel like not being in full control of our own directory hierarchy opens the door to potential user configuration problems  we haven't faced yet.

Reply:
> I question the need for this field, because currently cachi2 cannot be configured like this for any pkg manager apart from pointing it to an alternative top-level directory.

In my opinion that is the case because for pkg manager, you generally don't care where the files are, as long as they are "available" in some way for the package manager during the build itself.

However this is not true for `generic` files, because they are often directly referenced in the Dockerfile. I will for now omit migration to cachi2 from OSBS (although important for the Konflux use case). However consider the following use case. 

Say you reference a versioned file (e.g. OVAL feed) in your Dockerfile. When the time comes to update to a new version, you will need to update the cachi2 lockfile (or the SBOM, whatever we decide to go with), AND update the Dockerfile, to change the reference. 

However,  with the `target` feature, you control the dependency only through the lockfile/SBOM, because updating the version of the dependency does not affect the way you use the dependency in the build, which is closer to the general experience when using a package manager. 
 
> I feel like not being in full control of our own directory hierarchy opens the door to potential user configuration problems we haven't faced yet.

I feel like this is an implementation detail that can be solved. As mentioned, cachi2 can ensure the path is always a subdirectory of `cachi2-output/deps/generic`

Reply:
> Say you reference a versioned file (e.g. OVAL feed) in your Dockerfile. When the time comes to update to a new version, you will need to update the cachi2 lockfile (or the SBOM, whatever we decide to go with), AND update the Dockerfile, to change the reference.
> 
> However, with the target feature, you control the dependency only through the lockfile/SBOM, because updating the version of the dependency does not affect the way you use the dependency in the build, which is closer to the general experience when using a package manager.

I don't get this part. Can you follow with an example? If you reference a versioned file in your Dockerfile, then changing the reference WILL require a change to the Dockerfile as well. I don't get how target rids you off the need to change the Dockerfile accordingly. If references are the issue and you really really don't want to update your Dockerfile, then you can install as `<cmd> <cachi2-output-path>/<mydependency>*.oval`

Reply:
I meant that in the Dockerfile, you don't care about the version, but about having the dependency there. 

An example might be that before using cachi2, you have this kind of stuff in the dockerfile.

```
RUN curl -o feed.oval "https://foo.com/bar-08-24.oval"

RUN somecommand feed.oval
```

Now say you want to start doing hermetic builds, therefore you need to pre-fetch that oval file. You will specify the url in lockfile/SBOM and cachi2 will pre-fetch it for you. Now when you want update the version of the file, you need to:
- update the lockfile/SBOM
- update the dockerfile from `RUN somecommand cachi2-output/deps/generic/bar-08-24.oval` to  `RUN somecommand cachi2-output/deps/generic/bar-09-24.oval`

Having the ability to specify target makes this easier. Given `target: feed.oval` you only need to need to change the lockfile/SBOM, because the dockerfile references the target, e.g.  `RUN somecommand cachi2-output/deps/generic/feed.oval`

Reply:
Okay, very well, instead of `<cmd> cachi2-output/generic/feed.oval` you can do `<cmd> cachi2-ouptut/generic/feed*.oval` so I don't see the technical need justifying this field, but if others are in favour, it's not a blocker for me having it. FWIW this doesn't go against a different input file format other than you're proposing.

Reply:
Thanks, closing this conversation as resolved then. I didn't include it as necessary, but as convenience for users.